4337/test/utils/webauthn.ts

```diff
@@ -0,0 +1,244 @@
+/**
+ * This module provides a minimal shim to emulate the Web Authentication API implemented in browsers. This allows us to
+ * write tests where we create and authenticate WebAuthn credentials that are verified on-chain.
+ *
+ * This implementation is inspired by software authenticators found in the Awesome WebAuthn list [1].
+ *
+ * [1]: <https://github.com/herrjemand/awesome-webauthn#software-authenticators>
+ */
+
+import { p256 } from '@noble/curves/p256'
+import { ethers, BytesLike } from 'ethers'
+import CBOR from 'cbor'
+import { isArrayBuffer } from 'util/types'
+
+export interface CredentialCreationOptions {
+  publicKey: PublicKeyCredentialCreationOptions
+}
+
+/**
+ * Public key credetial creation options, restricted to a subset of options that this module supports.
+ * See <https://w3c.github.io/webauthn/#dictionary-makecredentialoptions>.
+ */
+export interface PublicKeyCredentialCreationOptions {
+  rp: { id: string; name: string }
+  user: { id: Uint8Array; displayName: string; name: string }
+  challenge: Uint8Array
+  pubKeyCredParams: {
+    type: 'public-key'
+    alg: number
+  }[]
+  attestation?: 'none'
+}
+
+export interface CredentialRequestOptions {
+  publicKey: PublicKeyCredentialRequestOptions
+}
+
+/**
+ * Public key credetial request options, restricted to a subset of options that this module supports.
+ * See <https://w3c.github.io/webauthn/#dictionary-assertion-options>.
+ */
+export interface PublicKeyCredentialRequestOptions {
+  challenge: Uint8Array
+  rpId: string
+  allowCredentials: {
+    type: 'public-key'
+    id: Uint8Array
+  }[]
+  attestation?: 'none'
+}
+
+/**
+ * A created public key credential. See <https://w3c.github.io/webauthn/#iface-pkcredential>.
+ */
+export interface PublicKeyCredential<AuthenticatorResponse> {
+  type: 'public-key'
+  id: string
+  rawId: ArrayBuffer
+  response: AuthenticatorResponse
+}
+
+/**
+ * The authenticator's response to a client’s request for the creation of a new public key credential.
+ * See <https://w3c.github.io/webauthn/#iface-authenticatorattestationresponse>.
+ */
+export interface AuthenticatorAttestationResponse {
+  clientDataJSON: ArrayBuffer
+  attestationObject: ArrayBuffer
+}
+
+/**
+ * The authenticator's response to a client’s request generation of a new authentication assertion given the WebAuthn Relying Party's challenge.
+ * See <https://w3c.github.io/webauthn/#iface-authenticatorassertionresponse>.
+ */
+export interface AuthenticatorAssertionResponse {
+  clientDataJSON: ArrayBuffer
+  authenticatorData: ArrayBuffer
+  signature: ArrayBuffer
+  userHandle: ArrayBuffer
+}
+
+class Credential {
+  public id: string
+  public pk: bigint
+
+  constructor(
+    public rp: string,
+    public user: Uint8Array,
+  ) {
+    this.pk = p256.utils.normPrivateKeyToScalar(p256.utils.randomPrivateKey())
+    this.id = ethers.dataSlice(ethers.keccak256(ethers.dataSlice(p256.getPublicKey(this.pk, false), 1)), 12)
+  }
+
+  public cosePublicKey(): string {
```

Comment:
Took me a while to find out what cose is; perhaps adding a comment together with the method definition is not a bad idea (I know that the actual implementation links the spec, but I was confused by the method name) https://datatracker.ietf.org/doc/html/rfc8152

Reply:
The spec is referenced below twice (for example here: https://github.com/safe-global/safe-modules/blob/26fdff1d5d252962e431fd63ece40fa8fb5d7518/4337/test/utils/webauthn.ts#L101), but I’ll add doc comments referencing the spec again as it wasn’t super clear that those are links to the spec.

Reply:
Added more context to the function documentation.

acf4235e71d0535c3b0b8cddeb4bb16ec62aff7b